### PR TITLE
support non-lambda policies with the “logs” subcommand

### PR DESCRIPTION
**Pull request by skurfer**
Implemented in the execution modes’ respective `get_logs()` methods.

It could use a few more tests for the S3 and CloudWatch implementations, but I wanted to get it out the door so you could have a look.

### CODE REVIEW

**Comment by kapilt** [comment]
failures appear to be from unrelated upstream botocore/boto3 changes :-(

**Comment by kapilt** [comment]
unrelated failure issues should be addressed with merge of trunk

File: c7n/policy.py [hunk 1]
```diff
@@ -207,6 +213,29 @@ def run(self, *args, **kw):
                 "ActionTime", time.time() - at, "Seconds", Scope="Policy")
             return resources
 
+    def get_logs(self, start, end):
+        log_source = self.policy.ctx.output
```
Comment (kapilt) [inline c7n/policy.py]: there's a third variant here thats a bit simpler, which is if log group was passed to run and logs, then we can query out from cwl

File: c7n/policy.py [hunk 2]
```diff
@@ -207,6 +213,29 @@ def run(self, *args, **kw):
...
+        log_gen = ()
+        if log_source.use_s3():
+            session = utils.local_session(self.policy.session_factory)
+            raw_entries = log_entries_from_s3(session, log_source, start)
+            # log files can be downloaded out of order, so sort on timestamp
+            # log_gen isn't really a generator once we do this, but oh well
+            log_gen = sorted(
+                normalized_log_entries(raw_entries),
```
Comment (kapilt) [inline c7n/policy.py]: part of the intent was that we could normalize on object keys rather than parse of entries

File: c7n/logs_support.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        key_prefix,
+        start.strftime('%Y/%m/%d/00'),
+        log_filename,
+    )
+    p = client.get_paginator('list_objects').paginate(
+        Bucket=output.bucket,
+        Prefix=key_prefix + '/',
+        Marker=marker,
+    )
+    with ThreadPoolExecutor(max_workers=20) as w:
```
Comment (kapilt) [inline c7n/logs_support.py]: thread count feels a bit high given the gil, would be nice to verify its still an improvement.

File: c7n/logs_support.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        for key_set in p:
+            if 'Contents' not in key_set:
```
Comment (kapilt) [inline c7n/logs_support.py]: one other stop case, if end date marker is reached, then there's no point to continuing iteration.

File: c7n/logs_support.py (hunk 3 above)
Comment (skurfer) [inline c7n/logs_support.py]: It probably is high. It was based on similar logic in `csvout.py`.

File: c7n/logs_support.py (hunk 4 above)
Comment (skurfer) [inline c7n/logs_support.py]: I’ve added a test to the list comprehension to take the end date into account.

**Comment by skurfer** [comment]
> i was hoping to go a little coarser on the time match when searching s3 outputs to the hr key prefix at max vs per line regex processing.

The per-line processing is just the gymnastics required to make sure log entries always look the same (whether they come from local files, S3, or CloudWatch), and has little to do with date/time filtering. The dates are taken into account when fetching files.

The per-entry date/time filtering happens at the end and is applied no matter where the entries came from. Otherwise, “start” and “end” would have no effect on the Lambda execution mode. (They didn’t before either, but now that we’re adding start/end, it seems like they should operate consistently.)

File: c7n/utils.py [hunk 5]
```diff
@@ -343,3 +344,11 @@ def _f(*args, **kw):
             raise
     functools.update_wrapper(_f, f)
     return _f
+
+
+def timestamp_from_string(date_text):
```
Comment (kapilt) [inline c7n/utils.py]: given specialized usage on this i think just move to log support mod and maybe do the * 1000 there as well.

File: c7n/logs_support.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    log.info('Fetched {} records across {} files'.format(
+        len(records),
+        key_count,
+    ))
+    return records
+
+
+def get_records(bucket, key, client):
+    result = client.get_object(Bucket=bucket, Key=key['Key'])
```
Comment (kapilt) [inline c7n/logs_support.py]: clients and sessions shouldn't  be used across threads concurrently.

File: c7n/policy.py [hunk 7]
```diff
@@ -207,6 +214,35 @@ def run(self, *args, **kw):
...
             return resources
 
+    def get_logs(self, start, end):
+        log_source = self.policy.ctx.output
+        log_gen = ()
+        if self.policy.options.log_group is not None:
+            session = utils.local_session(self.policy.session_factory)
+            log_gen = log_entries_from_group(
+                session,
+                self.policy.options.log_group,
```
Comment (kapilt) [inline c7n/policy.py]: this will need start and end as well.

File: c7n/logs_support.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    for entry in entries:
+        log_timestamp = entry.get('timestamp', 0)
+        if log_timestamp >= start and log_timestamp <= end:
+            yield entry
+
+
+def log_entries_from_s3(session, output, start, end):
+    client = session.client('s3')
+    key_prefix = output.key_prefix.strip('/')
+    local_tz = tz.tzlocal()
```
Comment (kapilt) [inline c7n/logs_support.py]: i forgot are we doing tz local on output or utc?

**Comment by kapilt** [comment]
so datetime on s3 is encoded in the key, re start/end filtering down to the hr which can be min granularity supported there.

re regex, not sure we really need to normalize the output, or why they would be different given its the same formatter on stdlib logging.  if different its preferrable to do that with the records from cw logs which are already diced and the same format.

File: c7n/utils.py (hunk 5 above)
Comment (skurfer) [inline c7n/utils.py]: Sounds reasonable.

File: c7n/policy.py (hunk 7 above)
Comment (skurfer) [inline c7n/policy.py]: To filter on the remote end, you mean?

This is mostly taken from `LambdaManager.logs()`. Should I fix it there, too?

File: c7n/logs_support.py (hunk 8 above)
Comment (skurfer) [inline c7n/logs_support.py]: Output is in local time, but that doesn’t matter so much here. This code is comparing the end date, which is a naive local time, with the entry’s UTC timestamp. So this local timezone lets us do the two things needed for comparison:

  * add a timezone to the end date
  * convert the remote entry’s UTC date to local

**Comment by skurfer** [comment]
> re regex, not sure we really need to normalize the output, or why they would be different…

When I say entries should “look the same”, I mean each should be a dictionary with `timestamp` and `message`, since that’s what `LambdaManager.logs()` returns. (Of course that isn’t set in stone, but it was the only one implemented previously, so I tried to match it.)

As for the specific format of the “message” part of the entry, I was basing it on some existing CloudWatch logs out there in the sandbox. I can see now that there is no consistent format for those really, so if you’d like, I can dumb that down and just take the message as-is. That probably only saves one line of code, though. We still need all the regex stuff to generate timestamps and identify multi-line entries.

File: c7n/logs_support.py (hunk 6 above)
Comment (skurfer) [inline c7n/logs_support.py]: Fixed.

File: c7n/policy.py (hunk 7 above)
Comment (skurfer) [inline c7n/policy.py]: This is done now. I went ahead and updated the LambdaManager incarnation as well.

File: c7n/logs_support.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        Prefix=key_prefix + '/',
+        Marker=marker,
+    )
+    with ThreadPoolExecutor(max_workers=20) as w:
+        for key_set in p:
+            if 'Contents' not in key_set:
+                continue
+            keys = [k for k in key_set['Contents']
+                    if k['Key'].endswith(log_filename)
+                    and k['LastModified'].astimezone(local_tz) < end]
```
Comment (kapilt) [inline c7n/logs_support.py]: rather than converting every record to local tz, why not convert the local time to utc for comparison, from a number of conversions as cost perspective, one conversion is better than many.

**Comment by skurfer** [comment]
Speaking of that date comparison, not having garnaat/placebo#50 is throwing a wrench into the testing. I’ll add what I can, though.

**Comment by skurfer** [comment]
At least I was able to test something. Poor coverage on the S3 side I’m afraid because of issues with Placebo. There’s the timezone issue mentioned earlier, and then I ran into a problem with it trying to JSON encode the Gzip data we get from `get_object()`. I suspect that will be fixed if garnaat/placebo#53 ever gets merged.

File: c7n/logs_support.py (hunk 4 above)
Comment (kapilt) [inline c7n/logs_support.py]: doing it in the list comprehension, means we don't stop key_set page iteration, so we'd still be iterating through other records in the bucket, even though we won't be using any of them.

**Comment by skurfer** [comment]
I’ve added a commit to address the concerns about iteration. Unfortunately, there’s nothing at the `key_set` level that can tell us we need to stop. You need to look at either `LastModified` or `Key` on each item inside `key_set['Contents']`.

The list comprehension is in two steps now because, the way it was, simply checking that there were no matches *could* mean we passed the end date, but it could also mean there were no log files in that set. Now, if we see log files but none of them pass the date check, we know we’re past the end and can stop.

Where this could fail (as in keep iterating) is if there are tons and tons of pages with no log file objects, but that doesn’t seem likely unless someone manually added a bunch of objects to the same folder Custodian is logging to. If you’re worried about it, an alternate approach is to see if the `Key` represents a time after the end, but that gets messy.

**Comment by kapilt** [comment]
its more that we have policies that have been running for over a year > 70k records, and this as is will chew through lots of objects that are not going to match, using list objects v2 with prefix and start key, and the knowledge that the date is encoded in the key should be utilized to avoid doing extra calls and processing.

**Comment by skurfer** [comment]
> this as is will chew through lots of objects that are not going to match

Well, one page at most, but with pages being up to 1000 objects, I suppose that’s worth trimming down further.

I’ll work on that and switch to `list_objects_v2()`. I’ll switch `report` to V2 while I’m at it.

**Comment by kapilt** [comment]
more than one page, since the policy output path is a prefix, you'll have to look at lots of pages and objs, that are pre date of interest. sounds good re listv2

**Comment by kapilt** [comment]
your right, its one more page at max, looked over again and saw the marker (in v2 its a starting key).

**Comment by kapilt** [comment]
wrt to the s3 coverage, it would need to go single threaded for sanity but even then potentially runs afoul of the placebo bugs. i've given up on upstream due to lack of response, and considering a fork to get a maintained core dependency (for skew already filed an issue for removing).

**Comment by kapilt** [comment]
thinking we should just merge this as is, and work on improvements incrementally against trunk.
